gax/src/main/java/com/google/api/gax/batching/NonBlockingSemaphore.java

```diff
@@ -35,33 +35,85 @@
 
 /** A {@link Semaphore64} that immediately returns with failure if permits are not available. */
 class NonBlockingSemaphore implements Semaphore64 {
-  private final AtomicLong currentPermits;
+  private AtomicLong availablePermits;
+  private AtomicLong limit;
 
   private static void checkNotNegative(long l) {
     Preconditions.checkArgument(l >= 0, "negative permits not allowed: %s", l);
   }
 
   NonBlockingSemaphore(long permits) {
     checkNotNegative(permits);
-    this.currentPermits = new AtomicLong(permits);
+    this.availablePermits = new AtomicLong(permits);
+    this.limit = new AtomicLong(permits);
   }
 
+  @Override
   public void release(long permits) {
     checkNotNegative(permits);
-    currentPermits.addAndGet(permits);
+    while (true) {
+      long old = availablePermits.get();
+      // TODO: throw exceptions when the permits overflow
+      if (availablePermits.compareAndSet(old, Math.min(old + permits, limit.get()))) {
+        return;
+      }
+    }
   }
 
+  @Override
   public boolean acquire(long permits) {
     checkNotNegative(permits);
-
-    for (; ; ) {
-      long old = currentPermits.get();
+    while (true) {
+      long old = availablePermits.get();
       if (old < permits) {
         return false;
       }
-      if (currentPermits.compareAndSet(old, old - permits)) {
+      if (availablePermits.compareAndSet(old, old - permits)) {
+        return true;
+      }
+    }
+  }
+
+  @Override
+  public boolean acquirePartial(long permits) {
+    checkNotNegative(permits);
+    // To allow individual oversized requests to be sent, clamp the requested permits to the maximum
+    // limit. This will allow individual large requests to be sent. Please note that this behavior
+    // will result in availablePermits going negative.
+    while (true) {
+      long old = availablePermits.get();
+      if (old < Math.min(limit.get(), permits)) {
+        return false;
+      }
+      if (availablePermits.compareAndSet(old, old - permits)) {
         return true;
       }
     }
   }
+
+  @Override
+  public void increasePermitLimit(long permits) {
+    checkNotNegative(permits);
+    availablePermits.addAndGet(permits);
```

Comment:
This looks suspicious. Even though both `availablePermits` and `limit` are atomic, their joint update here is not atomic (if they were under the same lock, it would be). Is it ok here?

Reply:
Very good point and I struggled with it a bit but I think it's ok: 

- If multiple threads are calling `increasePermitsLimit`, this is ok because updates to the variables are atomic and we just need to make sure the numbers add up.
- If it happens concurrently with `reduceLimit` (which shouldn't really happen in reality), we need to make sure limit don't go < 0, and the checks in `reducePermit` should handle that. 
- If `increasePermitLimit` happens concurrently with `acquire`, limit is not accessed / updated in `acquire`, and `availablePermits` is atomic, so it should be fine.
- If  `increasePermitLimit` happens concurrently with `acquirePartial`, `limit` is accessed to check if there's enough permits. If limit is updated after `if (old < Math.min(limit.get(), permits))`, it means either `availablePermits` was also updated midway so `if (availablePermits.compareAndSet(old, old - permits))` won't succeed and it'll loop one more time, or `availablePermits` is also updated and it passed / failed the check.